core/services/keeper/registry_synchronizer_process_logs.go

```diff
@@ -0,0 +1,152 @@
+package keeper
+
+import (
+	"sync"
+
+	"github.com/smartcontractkit/chainlink/core/internal/gethwrappers/generated/keeper_registry_wrapper"
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/services/log"
+	"github.com/smartcontractkit/chainlink/core/services/postgres"
+	"github.com/smartcontractkit/chainlink/core/store/models"
+)
+
+func (rs *RegistrySynchronizer) processLogs(head models.Head) {
+	wg := sync.WaitGroup{}
+	wg.Add(3)
+	go rs.handleSyncRegistryLogs(head, wg.Done)
+	go rs.handleUpkeepCanceledLogs(head, wg.Done)
+	go rs.handleUpkeepRegisteredLogs(head, wg.Done)
+	wg.Wait()
+}
+
+func (rs *RegistrySynchronizer) handleSyncRegistryLogs(head models.Head, done func()) {
+	defer done()
+	oldEnough := isOldEnoughConstructor(head, rs.minConfirmations)
+	for {
+		i := rs.mailRoom.mbSyncRegistry.RetrieveIf(oldEnough)
+		if i == nil {
+			return
+		}
+		broadcast, ok := i.(log.Broadcast)
+		if !ok {
+			logger.Errorf("RegistrySynchronizer: invariant violation, expected log.Broadcast but got %T", broadcast)
+			continue
+		}
+		was, err := broadcast.WasAlreadyConsumed()
+		if err != nil {
+			logger.Warn(err)
+			continue
+		}
+		if was {
+			continue
+		}
+		if !head.IsInChain(broadcast.RawLog().BlockHash) {
+			continue
+		}
+		_, err = rs.syncRegistry()
+		if err != nil {
+			logger.Error(err)
+			continue
+		}
+		err = broadcast.MarkConsumed()
+		logger.ErrorIf(err)
+	}
+}
+
+func (rs *RegistrySynchronizer) handleUpkeepCanceledLogs(head models.Head, done func()) {
+	defer done()
+	oldEnough := isOldEnoughConstructor(head, rs.minConfirmations)
+	for {
+		i := rs.mailRoom.mbUpkeepCanceled.RetrieveIf(oldEnough)
+		if i == nil {
+			return
+		}
+		broadcast, ok := i.(log.Broadcast)
+		if !ok {
+			logger.Errorf("RegistrySynchronizer: invariant violation, expected log.Broadcast but got %T", broadcast)
+			continue
+		}
+		was, err := broadcast.WasAlreadyConsumed()
+		if err != nil {
+			logger.Warn(err)
+			continue
+		}
+		if was {
+			continue
+		}
+		if !head.IsInChain(broadcast.RawLog().BlockHash) {
+			continue
+		}
+		log, ok := broadcast.DecodedLog().(*keeper_registry_wrapper.KeeperRegistryUpkeepCanceled)
+		if !ok {
+			logger.Errorf("RegistrySynchronizer: invariant violation, expected UpkeepCanceled log but got %T", log)
+			continue
+		}
+		ctx, cancel := postgres.DefaultQueryCtx()
+		defer cancel()
```

Comment:
this will run only after the whole loop - is that what you intended?

Reply:
yes I think this is fine! See discussion here:
https://chainlink-core.slack.com/archives/C0117GGJB6Y/p1615992286023400

Reply:
yes I think this is fine!

Reply:
I believe it's a memory leak

Reply:
@jb the alternative is to do something like this...
```go
for {
  // ...
  ctx, cancel := postgres.DefaultQueryCtx()
  defer cancel()
  err = rs.orm.QueryWithCTX(ctx)
  cancel()
  // ...
}
```
I thought the readability tradeoff was worth it to just `defer cancel()`

Reply:
ship it :shipit:

Reply:
I think you'd want something like:

```go
for {
  // ...
  {
    ctx, cancel := postgres.DefaultQueryCtx()
    defer cancel()
    err = rs.orm.QueryWithCTX(ctx)
  }
  // ...
}
```

Reply:
@jb - I don't think this works as suggested. The alternative would be to introduce an anonymous function, but personally I think the readability justifies the memory hit.